File: leetcode/100-999/141. Linked List Cycle/solution_1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
public:
    bool hasCycle(ListNode *head)
    {
        set<ListNode *> memoryAddress;

        auto h = head;
        while (h)
        {
            if (find(memoryAddress.begin(), memoryAddress.end(), h) == memoryAddress.end())
                {
                    memoryAddress.emplace(h);
                    h = h->next;
                }
            else
                return true;
        }

        return false;
    }
};

File: leetcode/100-999/141. Linked List Cycle/solution_1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution
{
    public bool HasCycle(ListNode head)
    {
        HashSet<ListNode> memoryAddress = new HashSet<ListNode>();

        var h = head;
        while (h != null)
        {
            if (!memoryAddress.Contains(h))
            {
                memoryAddress.Add(h);
                h = h.next;
            }
            else
                return true;
        }

        return false;
    }
}